### PR TITLE
Add IamProxyAuthProvider.

### PR DESCRIPTION
**Pull request by eap**
This new provider uses client credentials to test permissions
on a project before granting access to a user. Thus, it proxies
the user's IAM permissions.

In all cases:
- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Comment by eap** [comment]
jenkins retest

**Comment by eap** [comment]
@rtitle, can I get info on the Jenkins failure here?

**Comment by rtitle** [comment]
jenkins retest

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+import org.broadinstitute.dsde.workbench.leonardo.model.{LeoAuthProvider, ServiceAccountProvider}
+import org.broadinstitute.dsde.workbench.model.{UserInfo, WorkbenchEmail}
+import org.broadinstitute.dsde.workbench.model.google.{GoogleProject, IamPermission}
+
+import scala.concurrent.duration.{FiniteDuration, MINUTES}
+import scala.concurrent.{ExecutionContext, Future}
+
+object IamProxyAuthProvider {
+  private sealed trait CacheKey
+  private case class ProjectAuthCacheKey(userEmail: WorkbenchEmail, userToken: OAuth2BearerToken, googleProject: GoogleProject, executionContext: ExecutionContext) extends CacheKey
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala]: Yeah good catch here -- breaking the `UserInfo` into a `(email, token)` for the cache key seems more correct. I can make a separate PR to do that for the SamAuthProvider.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        def load(key: CacheKey) = {
+          key match {
+            case ProjectAuthCacheKey(userEmail, userToken, googleProject, executionContext) =>
+              checkUserAccessFromIam(userEmail, userToken, googleProject)(executionContext)
+          }
+        }
+      }
+    )
+
+  protected def petGoogleIamDao(token: String)(implicit executionContext: ExecutionContext): GoogleIamDAO = {
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala]: Nit: this is not technically using the pet, but the user's token. So maybe name this `userGoogleIamDao`?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    * Leo calls this method when it receives a "list clusters" API call, passing in all non-deleted clusters from the database.
+    * It should return a list of clusters that the user can see according to their authz.
+    *
+    * @param userInfo The user in question
+    * @param clusters All non-deleted clusters from the database
+    * @return         Filtered list of clusters that the user is allowed to see
+    */
+  override def filterUserVisibleClusters(userInfo: UserInfo, clusters: List[(GoogleProject, ClusterName)])(implicit executionContext: ExecutionContext): Future[List[(GoogleProject, ClusterName)]] = {
+    // Check each project for user-access exactly once, then filter by project.
+    val projects = clusters.map(lv => lv._1).toSet.toList
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala]: Minor: the final `.toList` is probably not required

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala (hunk 1 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala]: Opened issue: https://github.com/DataBiosphere/leonardo/issues/451

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala (hunk 2 above)
Comment (eap) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala]: Good catch - fixed.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala (hunk 3 above)
Comment (eap) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala]: Fixed.

**Comment by eap** [comment]
jenkins retest

**Comment by rtitle** [comment]
Hey @eap, we're seeing a lot of test failures due to billing project creation in Rawls. It's not related to Leo, but is causing Leo test failures. We're investigating, stay tuned.

**Comment by rtitle** [comment]
jenkins retest

**Comment by rtitle** [comment]
As I’m out next week, please work with @kyuksel on getting this merged. I approved it; just needs a green build.

**Comment by eap** [comment]
jenkins retest

**Comment by kyuksel** [comment]
jenkins retest

**Pull request merged**